jetstream/core/orchestrator.py

```diff
@@ -508,6 +510,23 @@ def _prefill_thread(self, idx: int):
       del prefill_result
       del request
 
+  def _jax_transfer_prefill_result(self, new_request, target_idx):
+    new_request.prefill_result = jax.device_put(
+        new_request.prefill_result,
+        self._generate_engines[target_idx].get_prefix_destination_sharding(),
+    )
+    # Block here so we don't block on the generate thread that steps.
+    jax.block_until_ready(new_request.prefill_result)
+
+  def _ray_transfer_prefill_result(self, new_request, target_idx):
+    self._generate_engines[target_idx].transfer(new_request.prefill_result)
```

Comment:
I don't see anything ray specfic here, the transfer code is abstracted into engine.transfer(). Can we use a generic name such as "non_jax_transfer" instead of "ray_transfer" here? 
Jetstream doesn't need to know if its Ray or some other mechanism used. Also, lets move this setting to server config for better control.

Reply:
Done for the server config part. 

The code here is interface level, the actually logic (or real implementation) is in engine side (pytorch or maxtext). The implementation logic would be: 1: gather prefill result from TPU chips in ray worker 2: transfer all gathered result from TPU to  CPU ram though PCIE in ray worker3: transfer prefill result from prefill server to decode server though DCN by ray head. 

Right now or in near future, I don't see we will introduce another transfer mechanism beside ray or jax, I feel name it as ray_transffer is clear and straightforward.

Reply:
Based on the implementation of `_ray_transfer_prefill_result`, could we assign the responsibility of implementing transfer to the generate engine?

What I mean by this is that within the orchestrator, we can generically use:

```
def _transfer_prefill_result(
      self, new_request: ActiveRequest, target_idx: int
  ):
   self._generate_engines[target_idx].transfer(new_request.prefill_result)
```

and so your Pathways engine needs to implement transfer (the `jax.device_put` logic) and your Ray engine needs to implement transfer with ray.remote primitives.

Or does this break some assumption? I think having JAX involved in the orchestrator is a somewhat leaky abstraction

Reply:
Agree. @JoeZijunZhou May have more insights, but based on my understanding, this could break Pathway.

Reply:
Synced with @vipannalla offline, we should removed jax dependencies in JetStream in long term. Current concerns is what is the impact to Pathway, we will figure out and make sure Patyway can adopt no jax in jetstream.